test/functional/mempool_sigoplimit.py

```diff
@@ -0,0 +1,125 @@
+           witness script in the witness stack and the given output script."""
+        # create P2WSH address and fund it via MiniWallet first
+        txid, vout = self.wallet.send_to(
+            from_node=self.nodes[0],
+            scriptPubKey=script_to_p2wsh_script(witness_script),
+            amount=1000000,
+        )
+
+        # create spending transaction
+        tx = CTransaction()
+        tx.vin = [CTxIn(COutPoint(int(txid, 16), vout))]
+        tx.wit.vtxinwit = [CTxInWitness()]
+        tx.wit.vtxinwit[0].scriptWitness.stack = [bytes(witness_script)]
+        tx.vout = [CTxOut(500000, output_script)]
+        return tx
+
+    def test_sigops_limit(self, bytes_per_sigop, num_sigops):
+        sigop_equivalent_vsize = ceil(num_sigops * bytes_per_sigop / WITNESS_SCALE_FACTOR)
+        self.log.info(f"- {num_sigops} sigops (equivalent size of {sigop_equivalent_vsize} vbytes)")
+
+        # create a template tx with the specified sigop cost in the witness script
+        # (note that the sigops count even though being in a branch that's not executed)
+        num_multisigops = num_sigops // 20
+        num_singlesigops = num_sigops % 20
+        witness_script = CScript(
+            [OP_FALSE, OP_IF] +
+            [OP_CHECKMULTISIG]*num_multisigops +
+            [OP_CHECKSIG]*num_singlesigops +
+            [OP_ENDIF, OP_TRUE]
+        )
+        # use a 256-byte data-push as lower bound in the output script, in order
+        # to avoid having to compensate for tx size changes caused by varying
+        # length serialization sizes (both for scriptPubKey and data-push lengths)
+        tx = self.create_p2wsh_spending_tx(witness_script, CScript([OP_RETURN, b'X'*256]))
+
+        # bump the tx to reach the sigop-limit equivalent size by padding the datacarrier output
+        assert_greater_than_or_equal(sigop_equivalent_vsize, tx.get_vsize())
+        vsize_to_pad = sigop_equivalent_vsize - tx.get_vsize()
+        tx.vout[0].scriptPubKey = CScript([OP_RETURN, b'X'*(256+vsize_to_pad)])
+        assert_equal(sigop_equivalent_vsize, tx.get_vsize())
+
+        res = self.nodes[0].testmempoolaccept([tx.serialize().hex()])[0]
+        assert_equal(res['allowed'], True)
+        assert_equal(res['vsize'], sigop_equivalent_vsize)
+
+        # increase the tx's vsize to be right above the sigop-limit equivalent size
+        # => tx's vsize in mempool should also grow accordingly
+        tx.vout[0].scriptPubKey = CScript([OP_RETURN, b'X'*(256+vsize_to_pad+1)])
+        res = self.nodes[0].testmempoolaccept([tx.serialize().hex()])[0]
+        assert_equal(res['allowed'], True)
+        assert_equal(res['vsize'], sigop_equivalent_vsize+1)
+
+        # decrease the tx's vsize to be right below the sigop-limit equivalent size
+        # => tx's vsize in mempool should stick at the sigop-limit equivalent
+        # bytes level (i.e. exceeding the sigop limit is possible, but has to
+        # be compensated by higher fees)
```

Comment:
Not sure if I'm misunderstanding this comment but afaik the reason is that txsize is max(serialized size, "sigop" size), not that fees have any impact?

Reply:
The intention was to express the implication of exceeding the sig-op limit from a user's perspective (tx is treated as larger than it's serialized vsize => more fees needed to reach the same fee-rate; or as PR desc of #8365 put it "high-sigop transactions [...] need to pay a fee corresponding to the maximally-used resource."). Agree that the comment is rather misleading and not very helpful for the test, replaced it.

Generally struggled a bit with terminology for this PR (not even sure if talking about a "limit" is proper, usually we reject txs from mempool if limits are exceeded), suggestions of course very welcome.